packages/captp/src/atomics.js

```diff
@@ -0,0 +1,149 @@
+  const overheadLength = lenbuf.byteLength + statusbuf.byteLength;
+  assert.equal(
+    overheadLength,
+    TRANSFER_OVERHEAD_LENGTH,
+    X`Internal error; actual overhead ${overheadLength} of bytes is not TRANSFER_OVERHEAD_LENGTH ${TRANSFER_OVERHEAD_LENGTH}`,
+  );
+  const databuf = new Uint8Array(
+    transferBuffer,
+    lenbuf.byteLength + statusbuf.byteLength,
+  );
+  assert(
+    databuf.byteLength >= MIN_DATA_BUFFER_LENGTH,
+    X`Transfer buffer of size ${transferBuffer.byteLength} only supports ${databuf.byteLength} data bytes; need at least ${MIN_DATA_BUFFER_LENGTH}`,
+  );
+  return { statusbuf, lenbuf, databuf };
+};
+
+/**
+ * Create a trapHost that can be paired with makeAtomicsTrapGuest.
+ *
+ * This host encodes the transfer buffer and returns it in consecutive slices
+ * when the guest iterates over it.
+ *
+ * @param {SharedArrayBuffer} transferBuffer
+ * @returns {TrapHost}
+ */
+export const makeAtomicsTrapHost = transferBuffer => {
+  const { statusbuf, lenbuf, databuf } = splitTransferBuffer(transferBuffer);
+
+  const te = new TextEncoder();
+
+  return async function* trapHost([isReject, serialized]) {
+    // Get the complete encoded message buffer.
+    const json = JSON.stringify(serialized);
+    const encoded = te.encode(json);
+
+    // Send chunks in the data transfer buffer.
+    let i = 0;
+    let done = false;
+    while (!done) {
+      // Copy the next slice of the encoded arry to the data buffer.
+      const subenc = encoded.subarray(i, i + databuf.length);
+      databuf.set(subenc);
+
+      // Save the length of the remaining data.
+      lenbuf[0] = BigInt(encoded.length - i);
+
+      // Calculate the next slice, and whether this is the last one.
+      i += subenc.length;
+      done = i >= encoded.length;
+
+      // Find bitflags to represent the rejected and finished state.
+      const rejectFlag = isReject ? STATUS_FLAG_REJECT : 0;
+      const doneFlag = done ? STATUS_FLAG_DONE : 0;
+
+      // Notify our guest for this data buffer.
+
+      // eslint-disable-next-line no-bitwise
+      statusbuf[0] = rejectFlag | doneFlag;
+      Atomics.notify(statusbuf, 0, +Infinity);
+
+      // Wait until the next call.
+      yield;
+    }
+  };
+};
+
+/**
+ * Create a trapGuest that can be paired with makeAtomicsTrapHost.
+ *
+ * This guest iterates through the consecutive slices of the JSON-encoded data,
+ * then returns it.
+ *
+ * @param {SharedArrayBuffer} transferBuffer
+ * @returns {TrapGuest}
+ */
+export const makeAtomicsTrapGuest = transferBuffer => {
+  const { statusbuf, lenbuf, databuf } = splitTransferBuffer(transferBuffer);
+
+  return ({ startTrap }) => {
+    const td = new TextDecoder('utf-8');
+
+    let json = '';
+
+    // Start by sending the trap call to the host.
+    const it = startTrap();
+
+    let done = false;
+    while (!done) {
+      // Tell that we are ready for another buffer.
+      statusbuf[0] = STATUS_WAITING;
+      it.next();
```

Comment:
There is no way for captp to provide a synchronous iterator that has a legitimate `done`, as getting that value requires communication with the async trapHost.  What should I do in this case?  Add useless code?

Reply:
I've at least added an assertion.

Reply:
> There is no way for captp to provide a synchronous iterator that has a legitimate done, as getting that value requires communication with the async trapHost. What should I do in this case? Add useless code?

No. This is an example where my review comment gave bad advice because I misunderstood. Adding something to make things clearer, so that others reading the code don't fall into the same misunderstanding, would be great. That assertion may do. I'll look in a moment.

Following my bad advice, such as adding useless code, would have been terrible. Thanks for correcting me instead.

Reply:
The new code does indeed not suggest that misunderstanding. Resolving.